egs/wsj/s5/steps/nnet3/chain/get_egs.py

```diff
@@ -0,0 +1,727 @@
+        distances = []
+        additional_cost = 10
+        for cw in chunk_widths:
+            distance_from_left_boundary = width % cw
+            distance_from_right_boundary = cw - distance_from_left_boundary
+            if distance_from_left_boundary > distance_from_right_boundary:
+                distance = distance_from_right_boundary
+            else:
+                distance = distance_from_left_boundary + additional_cost
+            distances.append([cw, distance])
+        distances = sorted(distances, key = lambda x:x[1])
+        utt_partition_preference[utt_name] = map(lambda x: x[0], distances)
+
+    utts_per_chunk_width = {}
+    for cw in chunk_widths:
+        utts_per_chunk_width[cw] = []
+
+    # get the initial assignment
+    for utt, preferences in utt_partition_preference.items():
+        utts_per_chunk_width[preferences.pop(0)].append(utt)
+
+    # merge partitions until each of them has atleast frames_per_iter frames
+    for iteration in xrange(len(chunk_widths) - 1):
+        has_min_num_frames = True
+        for chunk_width in chunk_widths:
+            if GetNumFrames(feat_dir, utts_per_chunk_width[chunk_width]) <= frames_per_iter:
+                has_min_num_frames = False
+                # we don't want to use this chunk_width if there are less
+                # than frames_per_iter frames as there will not be sufficient
+                # randomization
+                for utt in utts_per_chunk_width.pop(chunk_width):
+                    preferences = utt_partition_preference[utt]
+                    while len(preferences) > 0:
+                        key = preferences.pop(0)
+                        if utts_per_chunk_width.has_key(key):
+                            utts_per_chunk_width[key].append(utt)
+                            break
+
+        if has_min_num_frames:
+            break
+    return utts_per_chunk_width
+
+def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir,
+                                     utts, chunk_width,
+                                     train_feats_string, train_egs_opts_string,
+                                     supervision_opts, ivector_opts,
+                                     num_jobs, max_shuffle_jobs_run,
+                                     frames_per_iter, cmd, only_shuffle):
+
+    # The examples will go round-robin to egs_list.  Note: we omit the
+    # 'normalization.fst' argument while creating temporary egs: the phase of egs
+    # preparation that involves the normalization FST is quite CPU-intensive and
+    # it's more convenient to do it later, in the 'shuffle' stage.  Otherwise to
+    # make it efficient we need to use a large 'nj', like 40, and in that case
+    # there can be too many small files to deal with, because the total number of
+    # files is the product of 'nj' by 'num_archives_intermediate', which might be
+    # quite large.
+    num_frames = data_lib.GetNumFrames(feat_dir, utts)
```

Comment:
... actually scrub the 'proportionally' thing.
My proposal was that the user specifies the
{min-max}-{chunk-size,extra-{left,right}-context}, and the number of
distinct sizes (say --num-sizes), and you just vary them along the same
axis, so that the smallest size will have the
min-{chunk-size,extra-{left,right}-context} and the largeset size would
have the max-{chunk-size,extra-{left,right}-context}
Dan


On Mon, Apr 11, 2016 at 1:13 PM, Daniel Povey <dpovey@gmail.com> wrote:

>
> > +def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir,
>>
>> > +                                     utts, chunk_width,
>> > +                                     train_feats_string, train_egs_opts_string,
>> > +                                     supervision_opts, ivector_opts,
>> > +                                     num_jobs, max_shuffle_jobs_run,
>> > +                                     frames_per_iter, cmd, only_shuffle):
>> > +
>> > +    # The examples will go round-robin to egs_list.  Note: we omit the
>> > +    # 'normalization.fst' argument while creating temporary egs: the phase of egs
>> > +    # preparation that involves the normalization FST is quite CPU-intensive and
>> > +    # it's more convenient to do it later, in the 'shuffle' stage.  Otherwise to
>> > +    # make it efficient we need to use a large 'nj', like 40, and in that case
>> > +    # there can be too many small files to deal with, because the total number of
>> > +    # files is the product of 'nj' by 'num_archives_intermediate', which might be
>> > +    # quite large.
>> > +    num_frames = data_lib.GetNumFrames(feat_dir, utts)
>>
>> OK, I will modify the left and right contexts to ensure that the total
>> number of input frames is always constant.
>>
>
>
> Er- not sure what you mean by this exactly.  What I was proposing was that
> as you increase the chunk size, you also increase the left-context,
> proportionally.  It wouldn't make sense to always leave the total
> num-frames in the example the same-- what about models that didn't even
> need extra left-context?  Anyway, this would generate very excessive
> context sizes if the range of chunk sizes was quite large.
>
> Dan
>
>
>
>> … <#m_-631027123077136216_m_-1947976277950078432_>
>> --Vijay
>> On Mon, Apr 11, 2016 at 4:05 PM, Daniel Povey <notifications@github.com>
>> wrote: In egs/wsj/s5/steps/nnet3/chain/get_egs.py <#686 (comment)
>> <https://github.com/kaldi-asr/kaldi/pull/686#discussion_r59271543>>: >
>> +def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir, >
>> + utts, chunk_width, > + train_feats_string, train_egs_opts_string, > +
>> supervision_opts, ivector_opts, > + num_jobs, max_shuffle_jobs_run, > +
>> frames_per_iter, cmd, only_shuffle): > + > + # The examples will go
>> round-robin to egs_list. Note: we omit the > + # 'normalization.fst'
>> argument while creating temporary egs: the phase of egs > + # preparation
>> that involves the normalization FST is quite CPU-intensive and > + # it's
>> more convenient to do it later, in the 'shuffle' stage. Otherwise to > + #
>> make it efficient we need to use a large 'nj', like 40, and in that case >
>> + # there can be too many small files to deal with, because the total
>> number of > + # files is the product of 'nj' by
>> 'num_archives_intermediate', which might be > + # quite large. > +
>> num_frames = data_lib.GetNumFrames(feat_dir, utts) My feeling is that as
>> long as there is some kind of variety present, it's not going to be able to
>> learn where the edges are-- having it very finely randomized is not needed.
>> The reason why I prefer having this localized to the get_egs script is that
>> it helps keep the training script simple-- and anyway we have to deal with
>> multiple chunk lengths in the get_egs script, so why not deal with the
>> multiple contexts there too? Dan … <#m_5709551275589886984_> On Mon, Apr
>> 11, 2016 at 12:43 PM, Vijayaditya Peddinti < notifications@github.com>
>> wrote: In egs/wsj/s5/steps/nnet3/chain/get_egs.py <#686
>> <https://github.com/kaldi-asr/kaldi/pull/686> (comment) <#686 (comment)
>> <https://github.com/kaldi-asr/kaldi/pull/686#discussion_r59267969>>>: >
>> +def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir, >
>> + utts, chunk_width, > + train_feats_string, train_egs_opts_string, > +
>> supervision_opts, ivector_opts, > + num_jobs, max_shuffle_jobs_run, > +
>> frames_per_iter, cmd, only_shuffle): > + > + # The examples will go
>> round-robin to egs_list. Note: we omit the > + # 'normalization.fst'
>> argument while creating temporary egs: the phase of egs > + # preparation
>> that involves the normalization FST is quite CPU-intensive and > + # it's
>> more convenient to do it later, in the 'shuffle' stage. Otherwise to > + #
>> make it efficient we need to use a large 'nj', like 40, and in that case >
>> + # there can be too many small files to deal with, because the total
>> number of > + # files is the product of 'nj' by
>> 'num_archives_intermediate', which might be > + # quite large. > +
>> num_frames = data_lib.GetNumFrames(feat_dir, utts) I think it would be
>> better to just dump all egs with max-left-context/max-right-context and
>> just perturb these during training. This might reduce the sensitivity of
>> the model to the contexts. — You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub <
>> https://github.com/kaldi-asr/kaldi/pull/686/files/ee9df73d5347ff02db360f6c317f00a202a3f064#r59267969>
>> — You are receiving this because you authored the thread. Reply to this
>> email directly or view it on GitHub <
>> https://github.com/kaldi-asr/kaldi/pull/686/files/ee9df73d5347ff02db360f6c317f00a202a3f064#r59271543>
>>
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/686/files/ee9df73d5347ff02db360f6c317f00a202a3f064#r59272066>
>>
>
>